### PR TITLE
feat: Added delay button in the scaled down revision (#1355)

### PR DESCRIPTION
**Pull request by schakrad**
fix for https://github.com/argoproj/argo-rollouts/issues/1355

Before the addition of button.
Screen Shot 2022-02-07 at 11 52 30 PM

After the addition of button to show the scaleDownDelay in seconds
Screen Shot 2022-02-08 at 12 00 49 AM

Signed-off-by: “schakradari” <“saisindhu_chakradari@intuit.com”>

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-rollouts/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [ ] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed my commits with [DCO](https://github.com/argoproj/argoproj)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My builds are green. Try syncing with master if they are not. 
* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-rollouts/blob/master/USERS.md).

### CODE REVIEW

**Comment by harikrongali** [comment]
@schakrad can you cleanup the PR and have diff show only required changes

File: ui/package.json [hunk 1]
```diff
@@ -60,7 +60,7 @@
         "sass": "^1.32.8",
         "ts-loader": "^8.0.17",
         "webpack-bundle-analyzer": "^4.4.0",
-        "webpack-cli": "^4.5.0",
```
Comment (harikrongali) [inline ui/package.json:63]: is this change required?

File: ui/yarn.lock [hunk 2]
```diff
@@ -2256,22 +2256,22 @@
     "@webassemblyjs/wast-parser" "1.9.0"
     "@xtuc/long" "4.2.2"
 
-"@webpack-cli/configtest@^1.0.3":
```
Comment (harikrongali) [inline ui/yarn.lock:2259]: not sure if lock file needs to be committed. @alexmt is it fine to commit this file?

**Comment by harikrongali** [comment]
Please add before and after snapshots as a part of the PR description.

Please fix CI checks: DCO & build
For DCO, you can squash commits and sign off on the last one

File: ui/package.json (hunk 1 above)
Comment (rbreeze) [reply ui/package.json:63]: +1

File: ui/yarn.lock (hunk 2 above)
Comment (rbreeze) [reply ui/yarn.lock:2259]: @harikrongali Lock file does need to be committed, but we should avoid bumping versions in any of these packages if possible. @schakrad I don't see any code changes that should require updating package versions, can you confirm?

**Review by rbreeze — CHANGES_REQUESTED** [review]
Code changes look good, my only request is to avoid changing package versions

File: ui/yarn.lock (hunk 2 above)
Comment (harikrongali) [reply ui/yarn.lock:2259]: Thanks, @rbreeze for confirmation. @schakrad please revert the package.json update

Comment (schakrad) [reply ui/yarn.lock:2259]: Sure Hari, I will check and revert back.

**Review by harikrongali — APPROVED** [review]

File: ui/src/app/components/Timer/timer.tsx [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+           setTime(start);
+       }, 1000);   
+       return () => {
+         if(timer) {
+            clearInterval(timer);
+         }
+       }
+   }, [props.date]);
+   return (
+      time === 0  ? null: <ActionButton label={`delay: ${time}s`} />
```
Comment (alexmt) [inline ui/src/app/components/Timer/timer.tsx:25]: Looks like this component is not just Timer. It has a button and `delay` label so it should be renamed accordingly. 

The `argo-ui` has `Ticker` component. So might consider just using it and "inline" implementation into `pods.tsx`. E.g.: 


pods.tsx
```tsx
import {Ticker} from 'argo-ui';

...
  <Ticker>
  {now => {
    // calculate time here
    return time === 0 ? null: <ActionButton label={`delay: ${time}s`} />;
  }}
...
```

**Review by alexmt — CHANGES_REQUESTED** [review]

File: ui/src/app/components/Timer/timer.tsx (hunk 3 above)
Comment (schakrad) [reply ui/src/app/components/Timer/timer.tsx:25]: Thanks Alex, I will check and update.

File: ui/src/app/components/pods/pods.tsx [hunk 4]
```diff
@@ -107,6 +110,16 @@ export const ReplicaSet = (props: {rs: RolloutReplicaSetInfo; showRevision?: boo
                 <ThemeDiv className='pods__header'>
                     <span style={{marginRight: '5px'}}>{rsName}</span> <ReplicaSetStatusIcon status={props.rs.status as ReplicaSetStatus} />
                     {props.showRevision && <div style={{marginLeft: 'auto'}}>Revision {props.rs.revision}</div>}
+                    {props.rs.scaleDownDeadline && (
+                        <div style={{marginLeft: 'auto'}}>
+                            <Ticker>
+                                {(now) => {
+                                    const time = moment(props.rs.scaleDownDeadline).diff(now, 'second');
+                                    return time === 0 ? null : <ActionButton label={`delay: ${time}s`} />;
```
Comment (alexmt) [inline ui/src/app/components/pods/pods.tsx:118]: I don't think it makes sense to use `ActionButton` here because user cannot really click anything and trigger any action. Can we use info item instead:

image

revision.tsx
```typescript
const ImageItems = (props: {images: ImageInfo[]}) => {
    return (
        <div>
            {props.images.map((img) => {
                let imageItems = img?.tags?.map((t) => {
                    return {content: t, icon: IconForTag(t)} as InfoItemProps;
                });
                if (imageItems.length === 0) {
                    imageItems = [];
                }
                imageItems.push({content: <Tooltip content='Scaledown in 3 hours 5 mins'><span>3 hours 5 mins</span></Tooltip>, icon: 'fa fa-clock'} as any);
                return <InfoItemRow key={img.image} label={<ThemeDiv className={`image image--${img.color || 'unknown'}`}>{img.image}</ThemeDiv>} items={imageItems} />;
            })}
        </div>
    );
};

```

**Review by alexmt — CHANGES_REQUESTED** [review]

**Review by alexmt — APPROVED** [review]

**Comment by alexmt** [comment]
LGTM. 


@rbreeze can you please take a look one more time?

**Review by rbreeze — CHANGES_REQUESTED** [review]
Just one last comment, can we revert `package.json` and `yarn.lock` changes?

**Comment by alexmt** [comment]
@rbreeze changes in package.json are actually required. Apparently `argo-ui`/`argo-cd/ui` and `argo-rollouts/ui` were using different `classnames` versions. I temporary downgraded `classnames` to unblock this PR.

**Comment by rbreeze** [comment]
@alexmt Oh I see, thank you!

**Review by rbreeze — APPROVED** [review]
LGTM!
